server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/DefaultMetaRepository.java

```diff
@@ -16,142 +16,291 @@
-        try {
-            final int headRev = normalizeNow(Revision.HEAD).major();
-            final Set<String> repos = parent().repos().list().keySet();
-            if (headRev > mirrorRev || !mirrorRepos.equals(repos)) {
-                mirrors = loadMirrors(headRev);
-                mirrorRev = headRev;
-                mirrorRepos = repos;
+    public CompletableFuture<List<Mirror>> mirrors(boolean includeDisabled) {
+        if (includeDisabled) {
+            return allMirrors();
+        }
+        return allMirrors().thenApply(mirrors -> {
+            return mirrors.stream().filter(Mirror::enabled).collect(toImmutableList());
+        });
+    }
+
+    @Override
+    public CompletableFuture<Mirror> mirror(String id) {
+        final String mirrorFile = mirrorFile(id);
+        return find(mirrorFile).thenCompose(entries -> {
+            @SuppressWarnings("unchecked")
+            final Entry<JsonNode> entry = (Entry<JsonNode>) entries.get(mirrorFile);
+            if (entry == null) {
+                throw new EntryNotFoundException("failed to find credential '" + mirrorFile + "' in " +
+                                                 parent().name() + '/' + name());
             }
 
-            return mirrors;
-        } finally {
-            mirrorLock.unlock();
+            return credentials().thenApply(credentials -> {
+                try {
+                    return parseMirror(mirrorFile, entry, credentials);
+                } catch (Exception e) {
+                    throw new RepositoryMetadataException("failed to load the credential configuration", e);
+                }
+            });
+        });
+    }
+
+    @Override
+    public CompletableFuture<Revision> saveMirror(MirrorDto mirrorDto, Author author) {
+        validateMirror(mirrorDto);
+        final String summary;
+        if (MirrorDirection.valueOf(mirrorDto.direction()) == MirrorDirection.REMOTE_TO_LOCAL) {
+            summary = "Create a new mirror from " + mirrorDto.remoteUrl() + mirrorDto.remotePath() + '#' +
+                      mirrorDto.remoteBranch() + " into " + mirrorDto.localRepo() + mirrorDto.localPath();
+        } else {
+            summary = "Create a new mirror from " + mirrorDto.localRepo() + mirrorDto.localPath() + " into " +
+                      mirrorDto.remoteUrl() + mirrorDto.remotePath() + '#' + mirrorDto.remoteBranch();
         }
+        return commitMirror(mirrorDto, author, summary);
     }
 
-    private Set<Mirror> loadMirrors(int rev) {
-        // TODO(trustin): Asynchronization
-        final Map<String, Entry<?>> entries =
-                find(new Revision(rev), PATH_CREDENTIALS_AND_MIRRORS, Collections.emptyMap()).join();
+    @Override
+    public CompletableFuture<Revision> updateMirror(MirrorDto mirrorDto, Author author) {
+        validateMirror(mirrorDto);
+        final String summary = "Update the mirror '" + mirrorDto.id() + '\'';
+        return commitMirror(mirrorDto, author, summary);
```

Comment:
It would be not a good idea to change the ID of a mirror. I checked if the mirror exists before committing although that is not an atomic operation.